enterprise/ssl-impl/src/main/java/io/crate/protocols/ssl/SslContextProviderService.java

```diff
@@ -0,0 +1,99 @@
+ * Features, you represent and warrant that you have a valid Enterprise or
+ * Subscription Agreement with Crate.io.  Your use of the Enterprise Features
+ * if governed by the terms and conditions of your Enterprise or Subscription
+ * Agreement with Crate.io.
+ */
+
+package io.crate.protocols.ssl;
+
+import org.elasticsearch.common.component.AbstractLifecycleComponent;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.Scheduler;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+
+import static java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY;
+
+@Singleton
+public class SslContextProviderService extends AbstractLifecycleComponent {
+
+    private static final TimeValue FILES_WATCH_POLL_INTERVAL = TimeValue.timeValueSeconds(5);
+
+    private final SslContextProvider sslContextProvider;
+    private final Settings settings;
+    private final ThreadPool threadPool;
+
+    private volatile FilesWatcher filesWatcher;
+    private volatile Scheduler.Cancellable watchRoutine;
+
+
+    @Inject
+    public SslContextProviderService(Settings settings,
+                                     ThreadPool threadPool,
+                                     SslContextProvider sslContextProvider) {
+        this.settings = settings;
+        this.threadPool = threadPool;
+        this.sslContextProvider = sslContextProvider;
+    }
+
+    @Override
+    protected void doStart() {
+        try {
+            filesWatcher = new FilesWatcher();
+
+            var keystorePath = SslConfigSettings.SSL_KEYSTORE_FILEPATH.setting().get(settings);
+            if (!keystorePath.isEmpty()) {
+                var keyStoreWatcher = new FileWatcher(Paths.get(keystorePath), ENTRY_MODIFY);
+                keyStoreWatcher.addListener(event -> sslContextProvider.reloadSslContext());
+                filesWatcher.register(keyStoreWatcher);
+            }
+            var trustStorePath = SslConfigSettings.SSL_KEYSTORE_FILEPATH.setting().get(settings);
+            if (!trustStorePath.isEmpty()) {
+                var trustStoreWatcher = new FileWatcher(Paths.get(trustStorePath), ENTRY_MODIFY);
+                trustStoreWatcher.addListener(event -> sslContextProvider.reloadSslContext());
+                filesWatcher.register(trustStoreWatcher);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        watchRoutine = threadPool.scheduleWithFixedDelay(
+            filesWatcher,
+            FILES_WATCH_POLL_INTERVAL,
+            ThreadPool.Names.GENERIC);
+    }
+
+    @Override
+    protected void doStop() {
+        if (watchRoutine != null && !watchRoutine.isCancelled()) {
+            watchRoutine.cancel();
+        }
+        if (filesWatcher != null) {
```

Comment:
`IOUtils.closeWhileHandlingException();` could be used which takes care of the null check and IOException handling